internal/provider/data_source_gitlab_project_milestones.go

```diff
@@ -0,0 +1,129 @@
+		Schema: map[string]*schema.Schema{
+			"project_id": {
+				Description: "The ID or URL-encoded path of the project owned by the authenticated user.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"iids": {
+				Description: "Return only the milestones having the given `iid` (Note: ignored if `include_parent_milestones` is set as `true`).",
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeInt},
+				Optional:    true,
+			},
+			"state": {
+				Description:      "Return only `active` or `closed` milestones.",
+				Type:             schema.TypeString,
+				Optional:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice(validMilestoneStates, false)),
+			},
+			"title": {
+				Description: "Return only the milestones having the given `title`.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"search": {
+				Description: "Return only milestones with a title or description matching the provided string.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"include_parent_milestones": {
+				Description: "Include group milestones from parent group and its ancestors. Introduced in GitLab 13.4.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+			},
+			"milestones": {
+				Description: "List of milestones from a project.",
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: datasourceSchemaFromResourceSchema(gitlabProjectMilestoneGetSchema(), nil, nil),
+				},
+			},
+		},
+	}
+})
+
+func dataSourceGitlabProjectMilestonesRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	project := d.Get("project_id").(string)
+	options := gitlab.ListMilestonesOptions{
+		ListOptions: gitlab.ListOptions{
+			PerPage: 20,
+			Page:    1,
+		},
+	}
+
+	if v, ok := d.GetOk("iids"); ok {
+		options.IIDs = intSetToIntSlice(v.(*schema.Set))
+	}
+
+	if v, ok := d.GetOk("title"); ok {
+		options.Title = gitlab.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("state"); ok {
+		options.State = gitlab.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("search"); ok {
+		options.Search = gitlab.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("include_parent_milestones"); ok {
+		options.IncludeParentMilestones = gitlab.Bool(v.(bool))
+	}
+
+	optionsHash, err := hashstructure.Hash(&options, nil)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var milestones []*gitlab.Milestone
+	for options.Page != 0 {
+		paginatedMilestones, resp, err := client.Milestones.ListMilestones(project, &options, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		milestones = append(milestones, paginatedMilestones...)
+		options.Page = resp.NextPage
+	}
+
+	log.Printf("[DEBUG] get gitlab milestones from project: %s", project)
+	d.SetId(buildTwoPartIDInterface(project, optionsHash))
```

Comment:
```suggestion
	d.SetId(fmt.Sprintf("%s:%d", project, optionsHash))
```

even though I think that `project` isn't even required and `optionsHash` is enough ...